File: Java/src/Solution101_200/Solution110.java
package Solution101_200;

import tool.TreeNode;

/**
 * @author windows
 */
public class Solution110{
    public boolean isBalanced(TreeNode root) {
        if (root == null){
            return true;
        }
        boolean flag = isBalanced(root.left) && isBalanced(root.right);
        if (!flag){
            return false;
        }
        int left = dfs(root.left);
        int right = dfs(root.right);
        return Math.abs(left - right) <= 1;
    }

    private int dfs(TreeNode root){
        if (root == null){
            return 0;
        }
        int left = dfs(root.left) + 1;
        int right = dfs(root.right) + 1;
        return Math.max(left, right);
    }




    public static void main(String[] args) {
        Solution110 solution110 = new Solution110();
    }
}

File: C++/src/Solution101_200/Solution110.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution110.h"
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include <algorithm>
using namespace std;

bool Solution110::isBalanced(TreeNode *root) {
    return dfs(root) >= 0;
}

int Solution110::dfs(TreeNode *root){
    if (root  == nullptr){
        return 0;
    }
    int left  = dfs(root->left);
    int right = dfs(root->right);
    if (left == -1 || right == -1 || abs(left - right) > 1){
        return -1;
    }else{
        return max(left, right) + 1;
    }
}
